### PR TITLE
Add React Native diagnostics

### PR DESCRIPTION
**Pull request by aidenybai**
## Why

Callstack's latest React Native guide documents several performance and correctness patterns that React Doctor did not yet detect, including Reanimated 4 Worklets migrations, Bottom Sheet gesture coordination, Expo platform shaking, Babel plugin ordering, and Android release shrinking. Existing FlashList guidance also lagged its v2 API.

## What changed

- add seven source rules and three project-level checks derived from the guide
- capability-gate Expo SDK 54 and Reanimated 4 diagnostics
- use scope-aware binding and import resolution to avoid shadowing false positives
- strengthen FlashList v2 heterogeneous-list detection and refresh outdated recommendations
- consolidate JSX import provenance into a shared resolver used by React Native and Ink rules
- add liveness, regression, and capability tests plus a package changeset

## Local evaluation

| Check | Result |
| --- | --- |
| RDE roots completed | 197 |
| React Native-targeted roots | 99 |
| Repositories covered | 63 total / 52 React Native-targeted |
| Target diagnostics inspected | 22 |
| Confirmed true positives | 22 |
| Confirmed false positives | 0 |
| Strict fuzzing | 1,600 iterations across 8 changed source rules, no findings |

The target hits were 20 Reanimated scheduler migrations in software-mansion/react-native-gesture-handler and two Android release shrinking findings in public React Native applications.

## Test plan

- nr format
- nr test
- nr lint
- nr typecheck
- nr format:check
- nr smoke:json-report
- git diff --check
- targeted rule liveness and regression tests
- strict rule fuzzing
- local React Doctor Evals corpus inspection

## Daytona parity

Compared 2,000/2,000 pinned projects at base 416fdcc955e2a2f1ea06ebbf4e6caf71af736fcf and final production tree at head 70e7b3410c00ff3fd3e3cdc6d419e407614d1fc6, with 0 skipped: 1,073,666 diagnostics unchanged, 843 added, and 808 removed. The 808 added/removed memoization diagnostics pair exactly at the same source locations and reflect copy/identity updates. The 35 net-new React Native findings (25 Worklets schedulers, 7 heterogeneous recyclable lists, 2 Android release-shrinking checks, and 1 React Compiler plugin-order check) were all inspected at their pinned sources: 35 true positives, 0 false positives. Follow-up parity used the same pinned 2,000-project corpus: b44a23fc5 to 4cb5901b5 exposed three false-negative removals, b960c052f restored exactly those three (+3/-0), c69ad7b75 was corpus-neutral against b960c052f (+0/-0), and the final review fixes at 0cbc2a61a were corpus-neutral against c69ad7b75 (+0/-0, 1,074,509 unchanged). Head eaa06f86d adds only the formatter-required deletion of a trailing blank line in a test file; its production tree remains identical to the tested and paritied 0cbc2a61a state.

Final incremental parity for the selector-correlation follow-up compared b960c052f21e27bc5fb0959db8d71a42bd404c84 to c69ad7b751a7e626e01fda4fa459dd12bc05f1a8 across 2,000/2,000 pinned projects with 0 skipped: 1,074,509 diagnostics unchanged, 0 added, and 0 removed.

---

> [!NOTE]
> **Medium Risk**
> Large static-analysis changes to heterogeneous list detection and new config parsers could mis-report on unusual Babel/Gradle setups, though capability gating and extensive tests limit blast radius.
> 
> **Overview**
> Adds **React Native diagnostics** aligned with current Callstack guidance: **seven new lint rules** (Gorhom Bottom Sheet scroll/gesture/`onAnimate` issues, Reanimated 4 removed APIs / spring thresholds / `runOnUI` → Worklets scheduler, Expo SDK 54 **platform tree-shaking** via direct imports) plus **three project-level checks** wired into `checkReactNativeProject` (static **Babel plugin order** for React Compiler and Reanimated 4 Worklets, **Android release minify/shrink** flags in `android/app` Gradle).
> 
> **Capability gating** now exposes `expo:54` and `reanimated:4` (and tracks `reanimatedVersion` on `ProjectInfo`) so version-specific rules only run when dependencies parse cleanly.
> 
> **List/recycler** coverage expands to Legend List v3 entrypoints and `AnimatedFlashList`, skips `estimatedItemSize` where v3/v2 defaults apply, and **heavily extends** `rn-list-recyclable-without-types` to infer heterogeneous `renderItem` roots through same-file helpers, fragments, logical/conditional JSX, and import-provenance-aware resolution (with shadowing guards).
> 
> Copy updates soften **React Compiler manual memoization** and **JS bottom sheet** recommendations; fuzz/regression tests cover short-circuit JSX and related control-flow edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit eaa06f86dd41b8b7f3f2bf308294c44b03eaa711. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 1]
```diff
@@ -1,23 +1,106 @@
...
+  const unwrappedExpression = stripParenExpression(expression);
+  if (isNodeOfType(unwrappedExpression, "JSXElement")) {
+    const elementName = getJsxElementQualifiedName(unwrappedExpression.openingElement.name);
+    if (elementName) names.add(elementName);
+    return;
+  }
+  if (isNodeOfType(unwrappedExpression, "ConditionalExpression")) {
+    collectReturnedJsxRootNames(unwrappedExpression.consequent, names);
+    collectReturnedJsxRootNames(unwrappedExpression.alternate, names);
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2259]: Fixed in 2ffa854ef: logical JSX return analysis now traverses both operands of &&, ||, and ?? expressions, with regressions for all three operators.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/utils/resolve-reanimated-api-name.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  callExpression: EsTreeNodeOfType<"CallExpression">,
+  scopes: ScopeAnalysis,
+  supportedApiNames: ReadonlySet<string>,
+): string | null =>
+  resolveReanimatedApiNameFromExpression(
+    callExpression.callee,
+    scopes,
+    supportedApiNames,
+    new Set(),
+  );
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/utils/resolve-reanimated-api-name.ts]: Fixed in 2ffa854ef: the Reanimated wrapper now delegates to resolveImportedApiReference, preserves type-only import guards centrally, and adds destructuring-alias coverage.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 3]
```diff
@@ -1,23 +1,111 @@
...
+  if (
+    initializer &&
+    (isNodeOfType(initializer, "ArrowFunctionExpression") ||
+      isNodeOfType(initializer, "FunctionExpression") ||
+      isNodeOfType(initializer, "FunctionDeclaration"))
+  ) {
+    return initializer;
+  }
+  return null;
+};
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2288]: Fixed in 5d04d0173: renderItem resolution now reuses getTransparentReactCallbackWrapperArgument with scope analysis, with regressions for an imported React useCallback wrapper and a shadowed local useCallback.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 4]
```diff
@@ -1,23 +1,111 @@
...
+  return returnedRootNames.size > 1;
+};
+
+const isFlashListV2OrNewer = (context: RuleContext): boolean => {
+  const flashListMajorVersion = getReactDoctorNumberSetting(
+    context.settings,
+    "shopifyFlashListMajorVersion",
+  );
+  return flashListMajorVersion !== undefined && flashListMajorVersion >= FLASH_LIST_V2_MAJOR;
+};
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2349]: Fixed in 5d04d0173: extracted isFlashListV2OrNewer to a shared React Native utility consumed by both FlashList rules.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 5]
```diff
@@ -1,23 +1,119 @@
...
+  if (isNodeOfType(unwrappedExpression, "ConditionalExpression")) {
+    collectReturnedJsxRootNames(unwrappedExpression.consequent, names);
+    collectReturnedJsxRootNames(unwrappedExpression.alternate, names);
+    return;
+  }
+  if (isNodeOfType(unwrappedExpression, "LogicalExpression")) {
+    collectReturnedJsxRootNames(unwrappedExpression.left, names);
+    collectReturnedJsxRootNames(unwrappedExpression.right, names);
+  }
+};
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2260]: Fixed in 3d07e6312: fragment returns now get one deterministic rendered-root signature, with regressions for mixed roots, differing fragment-only shapes, stable multi-child fragments, one-child equivalence, nested/comment transparency, and dynamic children.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-bottom-sheet-no-state-in-on-animate.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        if (
+          !statePair ||
+          !statePair.declarator.init ||
+          !isReactApiCall(statePair.declarator.init, "useState", context.scopes, {
+            allowGlobalReactNamespace: true,
+            resolveNamedAliases: true,
+          })
+        ) {
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-bottom-sheet-no-state-in-on-animate.ts]: Fixed in 3d07e6312: removed the stricter duplicate useState provenance check and now trust resolveReactUseStatePair, with a regression for the accepted unbound bare useState path.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 7]
```diff
@@ -1,23 +1,158 @@
...
+  if (childRootNames.length === 1) return childRootNames[0];
+  return `fragment:${JSON.stringify(childRootNames)}`;
+};
+
+const collectReturnedJsxRootNames = (expression: EsTreeNode, names: Set<string>): void => {
+  const unwrappedExpression = stripParenExpression(expression);
+  if (isNodeOfType(unwrappedExpression, "JSXElement")) {
+    const elementName = getJsxElementQualifiedName(unwrappedExpression.openingElement.name);
+    if (elementName) names.add(elementName);
+    return;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2203]: Fixed in faf2b5e5e: proven React Fragment elements now reuse the canonical isJsxFragmentElement resolver and flatten through the same row-shape path as shorthand fragments. Added named/default/namespace equivalence, different-shape, lexical-shadow, and non-React-import regressions.

File: packages/core/src/checks/react-native/check-babel-plugin-order.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        category: "Performance",
+        message:
+          "This Reanimated 4 project still uses `react-native-reanimated/plugin`, which was renamed to the Worklets Babel plugin.",
+        help: "Replace it with `react-native-worklets/plugin` and keep that plugin last in the Babel `plugins` array.",
+      }),
+    ];
+  }
+
+  const workletsPluginIndex = pluginNames.indexOf(WORKLETS_PLUGIN_NAME);
+  if (workletsPluginIndex === -1 || workletsPluginIndex === pluginNames.length - 1) return [];
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/core/src/checks/react-native/check-babel-plugin-order.ts]: Fixed in 518bea5b0: the Worklets order check now evaluates the final occurrence with lastIndexOf. Regressions cover a duplicate final Worklets entry staying silent and a plugin after the final Worklets occurrence still reporting.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 9]
```diff
@@ -1,31 +1,175 @@
...
+): EsTreeNode | null => {
+  if (!isNodeOfType(attribute.value, "JSXExpressionContainer")) return null;
+  const expression = stripParenExpression(attribute.value.expression);
+  const directFunction = resolveFunctionFromInitializer(expression, null, scopes);
+  if (directFunction) return directFunction;
+  if (!isNodeOfType(expression, "Identifier")) return null;
+  const symbol = scopes.symbolFor(expression);
+  if (!symbol?.initializer) return null;
+  return resolveFunctionFromInitializer(symbol.initializer, symbol, scopes);
+};
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2304]: Fixed in 4a90e74c4 by reusing resolveExactLocalFunction for named declarations and aliases, while preserving the reassignment guard before the React callback-wrapper fallback. Added regressions for direct declarations, const aliases, homogeneous output, reassignment, lexical shadowing, and imported handlers. Focused and full tests, strict fuzzing, and exact-head 2,000-project parity all pass; the fix is corpus-neutral (+0/-0 versus the prior head).

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 10]
```diff
@@ -1,31 +1,179 @@
...
+  const expression = stripParenExpression(attribute.value.expression);
+  const directFunction = resolveFunctionFromInitializer(expression, null, scopes);
+  if (directFunction) return directFunction;
+  if (!isNodeOfType(expression, "Identifier")) return null;
+  const localFunction = resolveExactLocalFunction(expression, scopes);
+  if (localFunction) return localFunction;
+  const symbol = scopes.symbolFor(expression);
+  if (symbol?.kind === "function") return null;
+  if (!symbol?.initializer) return null;
+  return resolveFunctionFromInitializer(symbol.initializer, symbol, scopes);
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2303]: Fixed in fccb9f4bc by limiting the fallback resolver to immutable const bindings; mutable let/var handlers and reassigned function declarations remain conservatively unresolved. Added both heterogeneous-to-homogeneous and homogeneous-to-heterogeneous regressions for let and var. Full tests, typecheck, strict fuzzing, and exact-head 2,000-project parity pass; the fix is corpus-neutral (+0/-0 versus the prior head).

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-jsx-descendant-opening-elements.ts [hunk 11]
```diff
@@ -1,25 +1,68 @@
...
+    for (const element of expression.elements) {
+      if (element && !isNodeOfType(element, "SpreadElement")) {
+        appendDescendant(element, descendants, true);
+      }
+    }
+    return;
+  }
+  if (isNodeOfType(expression, "SequenceExpression")) {
+    for (const sequenceExpression of expression.expressions) {
+      appendDescendant(sequenceExpression, descendants, true);
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-jsx-descendant-opening-elements.ts]: Fixed in fccb9f4bc by reusing getFinalSequenceExpressionValue and traversing only the rendered final comma operand. Added Bottom Sheet regressions proving discarded earlier JSX is ignored and final JSX is still collected. Full tests, typecheck, strict fuzzing, and exact-head 2,000-project parity pass; the fix is corpus-neutral (+0/-0 versus the prior head).

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts [hunk 12]
```diff
@@ -1,31 +1,178 @@
...
+  }
+  if (isNodeOfType(unwrappedExpression, "ConditionalExpression")) {
+    collectReturnedJsxRootNames(unwrappedExpression.consequent, names, scopes);
+    collectReturnedJsxRootNames(unwrappedExpression.alternate, names, scopes);
+    return;
+  }
+  if (isNodeOfType(unwrappedExpression, "LogicalExpression")) {
+    collectReturnedJsxRootNames(unwrappedExpression.left, names, scopes);
+    collectReturnedJsxRootNames(unwrappedExpression.right, names, scopes);
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-list-recyclable-without-types.ts:2259]: Fixed in bf82b34bb: JSX-left logical expressions now follow JavaScript short-circuit semantics (`&&` collects the right root; `||`/`??` collect the left root), with regression tests and a strict fuzz corpus fixture for all three operators.

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-jsx-descendant-opening-elements.ts [hunk 13]
```diff
@@ -1,25 +1,68 @@
...
+  }
+  if (isNodeOfType(expression, "ConditionalExpression")) {
+    appendDescendant(expression.consequent, descendants, true);
+    appendDescendant(expression.alternate, descendants, true);
+    return;
+  }
+  if (isNodeOfType(expression, "LogicalExpression")) {
+    appendDescendant(expression.left, descendants, true);
+    appendDescendant(expression.right, descendants, true);
+    return;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-jsx-descendant-opening-elements.ts:55]: Fixed in fac19f6a4: logical JSX result branches are now selected by short-circuit semantics in a shared utility used by both descendant and root collectors. The helper also resolves sequence-expression final values. Added mirrored positive/negative Bottom Sheet tests and a strict fuzz regression fixture.

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-logical-expression-result-branches.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+export const getStaticLogicalExpressionResultBranches = (
+  expression: EsTreeNodeOfType<"LogicalExpression">,
+): ReadonlyArray<EsTreeNode> => {
+  const leftResult = getFinalSequenceExpressionValue(expression.left);
+  if (!isNodeOfType(leftResult, "JSXElement") && !isNodeOfType(leftResult, "JSXFragment")) {
+    return [expression.left, expression.right];
+  }
+  return expression.operator === "&&" ? [expression.right] : [leftResult];
+};
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-logical-expression-result-branches.ts:121]: Fixed in b44a23fc5: logical result analysis now recursively carries truthiness/nullishness path facts across left-associative chains, deduplicates branch states at every boundary to prevent exponential growth, and normalizes sequence final values at the FlashList collector entry. Added chained, mixed-operator, literal, sequence, and 24-operand boundedness regressions plus updated fuzz fixtures.

File: packages/core/src/checks/react-native/utils/read-android-release-shrinking.ts [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    } catch {
+      return null;
+    }
+
+    const blockDetectionContents = maskGradleStrings(
+      contents.replace(/(["'])release\1/g, RELEASE_STRING_MARKER),
+    );
+    const releaseBlocks = getReleaseBlocks(blockDetectionContents);
+    const [releaseBlock] = releaseBlocks;
+    if (releaseBlock === undefined || releaseBlocks.length !== 1) return null;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/core/src/checks/react-native/utils/read-android-release-shrinking.ts:176]: Verified as non-actionable and keeping the conservative behavior. Gradle defines a project build script as either build.gradle or build.gradle.kts; the Kotlin DSL migration guide renames the file rather than establishing an alternate-file fallback. If the selected script is unreadable or structurally ambiguous, reading the other extension can report against a file Gradle does not use. Existing dynamic and multiple-release-block regressions preserve the no-diagnostic result. References: https://docs.gradle.org/current/userguide/build_file_basics.html and https://developer.android.com/build/migrate-to-kotlin-dsl

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-jsx-descendant-opening-elements.ts [hunk 16]
```diff
@@ -1,25 +1,70 @@
...
+  }
+  if (!includeStaticExpressionBranches) return;
+  if (isNodeOfType(expression, "JSXExpressionContainer")) {
+    appendDescendant(expression.expression, descendants, true);
+    return;
+  }
+  if (isNodeOfType(expression, "ConditionalExpression")) {
+    appendDescendant(expression.consequent, descendants, true);
+    appendDescendant(expression.alternate, descendants, true);
+    return;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-jsx-descendant-opening-elements.ts:49]: Fixed in 0cbc2a61a. Static boolean ternary tests now traverse only the reachable JSX arm; dynamic tests continue traversing both. Added focused reachable, unreachable, sequence-test, and dynamic regressions plus a fuzz regression corpus fixture. The plugin suite passed 23,582 tests, strict targeted fuzz passed 1,000 iterations, and 2,000-project parity is neutral.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/utils/resolve-imported-recycler-name.ts [hunk 17]
```diff
@@ -29,28 +20,19 @@ const getJsxMemberObjectName = (node: EsTreeNode): string | null => {
...
-        localName === canonicalName &&
-        packageSources.some((packageSource) =>
-          isNamespaceImportFromModule(node, jsxMemberObjectName, packageSource),
-        );
-      if (isNamespaceMemberOfOwner) return canonicalName;
-      continue;
+    for (const packageSource of packageSources) {
+      if (resolveImportedJsxComponentName(node, packageSource, scopes) === canonicalName) {
+        return canonicalName;
+      }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/utils/resolve-imported-recycler-name.ts:34]: Verified as non-actionable. resolveImportedJsxComponentName does return default, but none of the mapped public recycler entrypoints has a default component export. Current Shopify and Legend package declarations and official docs expose FlashList, AnimatedFlashList, LegendList, AnimatedLegendList, KeyboardAwareLegendList, and KeyboardAvoidingLegendList as named exports; historical Shopify 1.x and Legend 2.x entrypoints are named too. The pre-refactor getImportedNameFromModule helper also explicitly returned null for default imports, so this is not a regression. Accepting default would be unsound for @shopify/flash-list because that source owns both FlashList and AnimatedFlashList and an arbitrary local default name cannot identify which canonical component it represents. If a supported entrypoint adds a real default component export later, the sound extension is explicit source-to-canonical metadata. References: https://shopify.github.io/flash-list/docs/usage/ and https://legendapp.com/open-source/list/v3/api/
